src/dagmc/DagMC.cpp

```diff
@@ -0,0 +1,774 @@
+#include "DagMC.hpp"
+
+#include <string>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <limits>
+#include <algorithm>
+#include <set>
+
+#include <ctype.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include <math.h>
+#ifndef M_PI  /* windows */
+# define M_PI 3.14159265358979323846
+#endif
+
+#define MB_OBB_TREE_TAG_NAME "OBB_TREE"
+#define FACETING_TOL_TAG_NAME "FACETING_TOL"
+
+namespace moab {
+
+/* Tolerance Summary
+
+   Facet Tolerance:
+   Maximum distance between continuous solid model surface and faceted surface.
+     Performance:  increasing tolerance increased performance (fewer triangles)
+     Robustness:   should not be affected
+     Knowledge:    user must understand how coarser faceting influences accuracy
+                   of results
+*/
+
+const bool counting = false; /* controls counts of ray casts and pt_in_vols */
+
+// Empty synonym map for DagMC::parse_metadata()
+const std::map<std::string, std::string> DagMC::no_synonyms;
+
+// DagMC Constructor
+DagMC::DagMC(Interface* mb_impl, double overlap_tolerance, double p_numerical_precision) {
+  moab_instance_created = false;
+  // if we arent handed a moab instance create one
+  if (NULL == mb_impl) {
+    mb_impl = new moab::Core();
+    moab_instance_created = true;
+  }
+
+  // set the internal moab pointer
+  MBI = mb_impl;
+
+  // make new GeomTopoTool and GeomQueryTool
+  GTT = new moab::GeomTopoTool(MBI, false);
+  GQT = new moab::GeomQueryTool(GTT, overlap_tolerance, p_numerical_precision);
+
+  // This is the correct place to uniquely define default values for the dagmc settings
+  defaultFacetingTolerance = .001;
+}
+
+// Destructor
+DagMC::~DagMC() {
+  // delete the GeomTopoTool and GeomQueryTool
+  delete GTT;
+  delete GQT;
+
+  // if we created the moab instance
+  // clear it
+  if (moab_instance_created) {
+    MBI->delete_mesh();
+    delete MBI;
+  }
+}
+
+// get the float verision of dagmc version string
+float DagMC::version(std::string* version_string) {
+  if (NULL != version_string)
+    *version_string = std::string("DagMC version ") + std::string(DAGMC_VERSION_STRING);
+  return DAGMC_VERSION;
+}
+
+unsigned int DagMC::interface_revision() {
+  unsigned int result = 0;
+  const char* interface_string = DAGMC_INTERFACE_REVISION;
+  if (strlen(interface_string) >= 5) {
+    // start looking for the revision number after "$Rev: "
+    result = strtol(interface_string + 5, NULL, 10);
```

Comment:
Error Handling: If `strtol` returns a `0`, it may have failed to parse the number successfully from the string.  Specifically, if `strtol` returns `0` and `errno` is set to `ERANGE`, it means the number failed to parse.  It may be worth it to check for this error state here.

[1] http://www.cplusplus.com/reference/cstdlib/strtol/